# Module 02 : Manipulation de données avec Pandas

## Fil conducteur

Tout au long de ce notebook, nous allons travailler sur une **question de recherche** (fictive) :

> **Est-ce que le télétravail contribue à une meilleure performance des entreprises ?**

Pour cela, nous allons utiliser le fichier `entreprises_panel.csv`, un jeu de données fictif de 80 entreprises suivies sur 5 ans (2019-2023), avec leur secteur, leurs effectifs, leur chiffre d'affaires, leur rentabilité et leur mode de travail.

Chaque notion `pandas` que vous apprendrez dans ce notebook servira **directement** à avancer sur cette question.

**Avant de commencer**, consultez le [README de ce module](README.md). Assurez-vous d'être à l'aise avec les notions du module 01 (listes, dictionnaires, boucles, fonctions).


## 1. Découverte du DataFrame

### 1.1. Qu'est-ce que `pandas` ?

`pandas` est une bibliothèque Python spécialisée dans **la manipulation et l'analyse des données**. Elle est particulièrement adaptée aux données organisées sous forme de tableaux, comme celles que l'on retrouve dans les fichiers Excel, CSV ou dans les bases de données. Imaginez `pandas` comme un tableau Excel, mais que vous pilotez avec du code plutôt qu'à la souris. C'est l'outil que vous utiliserez pour la quasi-totalité de vos traitements de données de recherche, quel que soit votre domaine en sciences de gestion. Elle permet notamment de :

- importer des fichiers CSV ou Excel ;
- explorer une base de données ;
- sélectionner et filtrer des observations ;
- traiter les valeurs manquantes ;
- créer ou modifier des variables ;
- regrouper et agréger des données ;
- calculer des statistiques descriptives ;
- préparer les données pour des analyses statistiques et des visualisations.

Pour utiliser `pandas`, il faut l'importer sous l'alias `pd`

```python
import pandas as pd
```

On peut ensuite utiliser ses principales fonctionnalités avec `pd`. Par exemple, pour créer un petit tableau :


In [1]:
import pandas as pd

donnees = {
    "age": [25, 32, 28],
    "satisfaction": [4, 3, 5]
}

df = pd.DataFrame(donnees)

`df` (la variable) contient alors un `DataFrame`, c'est-à-dire une structure de données tabulaire composée de lignes et de colonnes.

Avec Pandas, les données sont principalement organisées à l'aide de deux structures : les **Series** et les **DataFrames**. Une Series représente une colonne de données, tandis qu'un DataFrame représente un tableau composé de plusieurs colonnes.

### 1.2. Series

Une Series est une suite de valeurs organisée avec un index.

In [2]:
import pandas as pd

ages = pd.Series([25, 32, 28, 41])

On peut la représenter ainsi :

```
0    25
1    32
2    28
3    41
```

Les nombres à gauche constituent l'index, tandis que les valeurs à droite correspondent aux données. Une Series peut représenter une variable d'une enquête, par exemple l'âge des répondants.

### 1.3. DataFrame

Un DataFrame est un tableau de données composé de lignes et de colonnes. C'est l'équivalent d'une feuille Excel.

In [3]:
donnees = {
    "age": [25, 32, 28],
    "satisfaction": [4, 3, 5]
}

df = pd.DataFrame(donnees)

On obtient :

```
   age  satisfaction
0   25             4
1   32             3
2   28             5
```

Chaque colonne est une Series et l'ensemble des colonnes forme le DataFrame.

On peut donc retenir :

- **Series** = une colonne de tableau
- **DataFrame** = un tableau entier de données

> Vous verrez ces deux mots partout dans la documentation `pandas` et dans les messages d'erreur. Cette distinction est importante pour la suite du cours, car la plupart des opérations de manipulation et d'analyse de données avec Pandas se feront sur des DataFrames.


### 1.4. Charger un fichier CSV

Généralement, lorsque l'on travaille avec des données, elles sont stockées dans un fichier **CSV** (Comma Separated Values) ou **Excel**. La première des choses à faire avant de pouvoir manipuler et analyser des données  `pandas` consiste donc à les charger (importer). Pour cela, `pandas` dispose de nombreuses fonctions et méthodes qui permettent de charger efficacement des données dans des DataFrames à partir de fichiers CSV ou Excel.

Pour un fichier CSV, on utilise la fonction `read_csv()` qui va charger les données et les transformer en un `DataFrame` afin que nous puissions les manipuler et les analyser avec Python. Prenons l'exemple d'un jeu de données enregistré sous le nom "*donnees.csv*". On va commencer par importer `pandas`, puis définir une variable `df` dans laquelle stocker les données du fichier à partir de la fonction `read_csv()` :

```python
import pandas as pd

df = pd.read_csv("donnees.csv")
```

Le fichier que nous allons utiliser dans ce module se trouve dans le dossier `assets/datasets/`, à la racine du repo. Il s'agit du fichier `entreprises_panel.csv`.


In [4]:
import pandas as pd      # On commence par importer pandas

df = pd.read_csv("../../assets/datasets/entreprises_panel.csv")  # On charge le fichier entreprises_panel.csv avec la fonction read_csv() et on la stocke dans la variable df

# L'utilisation de "df" (abréviation de DataFrame) comme nom de variable est une convention très répandue quand il n'y a qu'un seul jeu de données principal dans le notebook.

type(df)   # Pour connaître le type de données de la variable df (un DataFrame)


pandas.DataFrame

### 1.5. Charger un fichier Excel

Pour charger un fichier Excel, on utilise la fonction `read_excel()` :

```python
df = pd.read_excel("donnees.xlsx")
```

On peut ensuite utiliser les mêmes méthodes qu'on utiliserait avec un fichier CSV.

---

> Cette étape de chargement de fichier permet de passer d'une base d'enquête, d'un fichier administratif ou d'un jeu de données secondaire à une base que Python peut exploiter pour les analyses suivantes.

---

### 1.6. Premiers réflexes sur un nouveau jeu de données

Après avoir chargé un fichier, il est nécessaire de vérifier que les données ont été correctement importées. Ainsi, avant de faire quoi que ce soit d'autre, vous devez prendre l'habitude d'**explorer les données** en utilisant les fonctions dédiées de `pandas`.

#### 1.6.1. `df.head()`


In [5]:
# .head() affiche les premières lignes (5 par défaut), pour une vérification rapide

df.head()


,entreprise_id,annee,secteur,region,annee_creation,teletravail,effectifs,chiffre_affaires_keur,rentabilite,satisfaction_rh
0,E001,2019,Industrie,Occitanie,2018,1,529,115830.2,0.1706,7.7
1,E001,2020,Industrie,Occitanie,2018,1,590,125110.5,0.1497,6.7
2,E001,2021,Industrie,Occitanie,2018,1,610,123760.0,0.1594,5.5
3,E001,2022,Industrie,Occitanie,2018,1,652,127441.6,0.1703,8.2
4,E001,2023,Industrie,Occitanie,2018,1,731,140627.0,0.0057,7.0


**Ce que vous devez observer ici** : chaque ligne correspond à *une entreprise, pour une année donnée* (on appelle ça des données de **panel**, on y reviendra au module 05). Notez que les 5 premières lignes concernent toutes la même entreprise (`E001`), sur 5 années différentes. C'est le jeu de données qui est trié ainsi, pas un hasard.

#### 1.6.2. `df.tail()`


In [6]:
# .tail() affiche les dernières lignes (5 par défaut)

df.tail()

,entreprise_id,annee,secteur,region,annee_creation,teletravail,effectifs,chiffre_affaires_keur,rentabilite,satisfaction_rh
399,E080,2019,Technologies,Hauts-de-France,2008,0,63,15307.4,0.1415,5.4
400,E080,2020,Technologies,Hauts-de-France,2008,0,61,13126.4,0.0640,NaN
401,E080,2021,Technologies,Hauts-de-France,2008,0,57,13995.9,0.0530,5.0
402,E080,2022,Technologies,Hauts-de-France,2008,0,57,15615.3,0.1162,6.3
403,E080,2023,Technologies,Hauts-de-France,2008,0,60,14633.5,0.0819,6.7


#### 1.6.3. `df.shape`

In [7]:
# .shape donne le nombre de lignes et de colonnes, sous la forme (lignes, colonnes)

df.shape


(404, 10)

> [!NOTE]
> Vous aurez remarqué que `df.shape` n'a pas de parenthèses, contrairement à `df.tail(), par exemple. Cela s'explique par le fait que dans Pandas `df.shape` est un attribut (une variable attachée à l'objet) qui stocke une information fixe, contrairement à `df.tail()`, `df.info`... qui sont des méthodes (fonctions).

#### 1.6.4. `df.info()`

In [8]:
# .info() donne un résumé technique : nom des colonnes, nombre de valeurs non manquantes et le type de chaque colonne

df.info()


<class 'pandas.DataFrame'>
RangeIndex: 404 entries, 0 to 403
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   entreprise_id          404 non-null    str    
 1   annee                  404 non-null    int64  
 2   secteur                404 non-null    str    
 3   region                 404 non-null    str    
 4   annee_creation         404 non-null    int64  
 5   teletravail            404 non-null    int64  
 6   effectifs              404 non-null    int64  
 7   chiffre_affaires_keur  404 non-null    float64
 8   rentabilite            404 non-null    float64
 9   satisfaction_rh        396 non-null    float64
dtypes: float64(3), int64(4), str(3)
memory usage: 31.7 KB


**Comment lire `.info()` ?**
- `RangeIndex: 404 entries` -> 404 lignes au total
- Pour chaque colonne : son nom, son nombre de valeurs **non nulles** (`Non-Null Count`), et son type (`Dtype`)
- Si `Non-Null Count` est inférieur au nombre total de lignes pour une colonne, cela signifie qu'il y a des **valeurs manquantes**. Regardez `satisfaction_rh` ; c'est justement le cas ici et on s'en occupera à la section 3.
- `Dtype` (le type de données) : `int64` pour des entiers, `float64` pour des décimaux, `str` pour du texte.


#### 1.6.5. `df.columns`

In [9]:
# .columns affiche les noms des colonnes

df.columns

Index(['entreprise_id', 'annee', 'secteur', 'region', 'annee_creation',
       'teletravail', 'effectifs', 'chiffre_affaires_keur', 'rentabilite',
       'satisfaction_rh'],
      dtype='str')

#### 1.6.6. `df.describe()`

In [10]:
# .describe() calcule des statistiques descriptives pour toutes les colonnes numériques

df.describe()


,annee,annee_creation,teletravail,effectifs,chiffre_affaires_keur,rentabilite,satisfaction_rh
count,404.000000,404.000000,404.000000,404.000000,404.000000,404.000000,396.000000
mean,2020.997525,2003.851485,0.574257,273.913366,41128.630941,0.086866,6.815404
std,1.420339,7.792133,0.495068,163.691548,27349.550234,0.063562,1.381483
min,2019.000000,1990.000000,0.000000,18.000000,2251.400000,-0.079100,3.000000
25%,2020.000000,1997.750000,0.000000,108.000000,14673.300000,0.042275,5.875000
50%,2021.000000,2003.000000,1.000000,271.500000,43071.100000,0.088550,6.800000
75%,2022.000000,2010.000000,1.000000,413.000000,60756.475000,0.128400,7.700000
max,2023.000000,2018.000000,1.000000,731.000000,140627.000000,0.270700,10.000000


**Pourquoi faire ça systématiquement ?** `.describe()` est en quelque sorte le meilleur moyen de repérer un problème dans vos données *avant* de vous lancer dans l'analyse :

- une valeur `min` négative pour un effectif (impossible) ;
- un `max` aberrant ;
- une moyenne qui ne colle pas avec ce que vous attendiez...

Prenez l'habitude de toujours vérifier ces chiffres avant d'aller plus loin.

**Rappel**

```python
df.head()             # premières lignes
df.tail()             # dernières lignes
df.shape              # nombre de lignes et de colonnes
df.columns            # noms des colonnes
df.info()             # informations générales
df.describe()         # calcul des statistiques descriptives pour toutes les colonnes numériques
```

---

> [!NOTE]
> Nous utilisons directement les fonctions `df.head()`, `df.tail()`... parce que nous sommes dans un notebook (en mode interactif). Cela permet d'avoir un beau tableau formaté. Mais dans un script Python `.py`, il faudra utiliser la fonction `print()`, car Python n'affiche rien sans print hors du mode interactif. Dans un script, vous devez alors faire `print(df.tail())`, par exemple, pour afficher les dernières lignes.

---

**À vous de jouer** : combien y a-t-il de lignes et de colonnes dans le DataFrame `df` de notre jeu de données ? Affichez uniquement ce chiffre (pas tout le DataFrame), en utilisant l'attribut vu plus haut.


In [11]:
# Votre code ici


## 2. Sélectionner et filtrer

Vous savez désormais comment charger des données à partir d'un fichier CSV ou Excel dans un DataFrame. Maintenant, il peut arriver que vous n'ayez besoin que d'une partie des données chargées. Dans ce cas, on va utiliser les options de **sélection** et de **filtrage**. La sélection consiste à choisir certaines lignes ou colonnes d'un DataFrame, tandis que le filtrage consiste à conserver uniquement les observations qui répondent à une condition.

### 2.1. Sélectionner une colonne

Pour extraire une seule colonne (une `Series` donc), on va utiliser des crochets avec le nom de la colonne entre guillemets :


In [12]:
df["secteur"] # On obtient alors une Series contenant uniquement la variable secteur.


0         Industrie
1         Industrie
2         Industrie
3         Industrie
4         Industrie
           ...     
399    Technologies
400    Technologies
401    Technologies
402    Technologies
403    Technologies
Name: secteur, Length: 404, dtype: str

Notez la ligne tout en bas de l'affichage : `Name: secteur`, `Length: 404`, `dtype: str`. C'est la signature d'une `Series`, pas d'un `DataFrame`. Si on veut sélectionner **plusieurs colonnes** (et donc obtenir un `DataFrame`, pas une `Series`), on va plutôt utiliser une **liste** de noms de colonnes entre doubles crochets :


In [13]:
df[["entreprise_id", "annee", "secteur", "teletravail"]]


,entreprise_id,annee,secteur,teletravail
0,E001,2019,Industrie,1
1,E001,2020,Industrie,1
2,E001,2021,Industrie,1
3,E001,2022,Industrie,1
4,E001,2023,Industrie,1
...,...,...,...,...
399,E080,2019,Technologies,0
400,E080,2020,Technologies,0
401,E080,2021,Technologies,0
402,E080,2022,Technologies,0


***Remarque sur la syntaxe :** le premier `[ ]` sert à "accéder à des colonnes", le second `[ ]` (à l'intérieur) est la liste Python des noms voulus ; d'où les doubles crochets `[[...]]`. Cela peut prêter à confusion au début.*

### 2.2. `.iloc` et `.loc` : sélectionner des lignes

Pour sélectionner précisément des données (des lignes) dans un DataFrame, `pandas` propose deux principales méthodes :

- **`.iloc`** : permet de sélectionner des données à partir de leur **position numérique** (comme un index de liste 0, 1, 2...) ;
- **`.loc`** : permet de sélectionner des données à partir de leurs **étiquettes** (noms d'index ou de colonnes).


In [14]:
# .iloc : "je veux la ligne à la position 0" (la toute première, peu importe son label)

df.iloc[0]


entreprise_id                 E001
annee                         2019
secteur                  Industrie
region                   Occitanie
annee_creation                2018
teletravail                      1
effectifs                      529
chiffre_affaires_keur     115830.2
rentabilite                 0.1706
satisfaction_rh                7.7
Name: 0, dtype: object

In [15]:
# .loc : ici, comme l'index de notre DataFrame est numéroté 0, 1, 2... par défaut
# (pas de colonne définie explicitement comme index), .loc[0] donne le même résultat
# que .iloc[0] -- mais ce n'est PAS toujours le cas, voir la remarque juste après

df.loc[0]


entreprise_id                 E001
annee                         2019
secteur                  Industrie
region                   Occitanie
annee_creation                2018
teletravail                      1
effectifs                      529
chiffre_affaires_keur     115830.2
rentabilite                 0.1706
satisfaction_rh                7.7
Name: 0, dtype: object

**Pourquoi cette distinction ?** Si vous filtrez ou triez votre DataFrame, les labels des lignes (l'index) ne correspondent plus à leur position. Exemple :


In [16]:
df_trie = df.sort_values("chiffre_affaires_keur", ascending=False)

# La ligne ayant le label (l'index) 42, quelle que soit sa position actuelle après le tri :
# df_trie.loc[42]

# La ligne qui est maintenant en première position après le tri, peu importe son label :

premiere_ligne_apres_tri = df_trie.iloc[0]
print(premiere_ligne_apres_tri[["entreprise_id", "annee", "chiffre_affaires_keur"]])


entreprise_id                E001
annee                        2023
chiffre_affaires_keur    140627.0
Name: 4, dtype: object


**Règle pratique** : Dans ce cours, nous utiliserons `.loc` presque tout le temps (pour sélectionner par condition, ce qu'on voit juste en dessous), et `.iloc` seulement quand nous voulons explicitement "la Nème ligne", indépendamment de son label.

### 2.3. Sélectionner lignes ET colonnes avec `.loc`

`.loc[lignes, colonnes]` 

> Remarquez la virgule qui sépare les deux :


In [17]:
# Toutes les lignes, seulement 3 colonnes précises

df.loc[:, ["entreprise_id", "annee", "chiffre_affaires_keur"]]


,entreprise_id,annee,chiffre_affaires_keur
0,E001,2019,115830.2
1,E001,2020,125110.5
2,E001,2021,123760.0
3,E001,2022,127441.6
4,E001,2023,140627.0
...,...,...,...
399,E080,2019,15307.4
400,E080,2020,13126.4
401,E080,2021,13995.9
402,E080,2022,15615.3


Le `:` seul signifie "toutes les lignes". C'est la même logique que le *slicing* sur les listes Python vu au module 01.

### 2.4. Filtrer selon une condition

C'est probablement l'opération que vous ferez le plus souvent avec `pandas` : garder seulement les lignes qui respectent une condition. Le filtrage est très courant en recherche. Il permet par exemple de :

- sélectionner les répondants d'une catégorie donnée ;
- étudier uniquement les entreprises d'un secteur ;
- analyser une période particulière ;
- exclure certaines observations ;
- travailler sur un sous-échantillon.

La mécanique se fait en deux temps, même si on l'écrit généralement en une seule ligne.


In [18]:
# Étape 1 : construire un "masque" -- une Series de True/False, une valeur par ligne

masque = df["teletravail"] == 1
masque.head()


0    True
1    True
2    True
3    True
4    True
Name: teletravail, dtype: bool

**Ce qui se passe ici** : `pandas` compare `df["teletravail"]` (une colonne entière) à `1`, ligne par ligne, et retourne une Series de `True`/`False` de la même longueur que `df`. C'est ce qu'on appelle un **masque booléen**.


In [19]:
# Étape 2 : utiliser ce masque pour filtrer le DataFrame -- il ne garde que les lignes où le masque vaut True

entreprises_teletravail = df[masque]
entreprises_teletravail.head()


,entreprise_id,annee,secteur,region,annee_creation,teletravail,effectifs,chiffre_affaires_keur,rentabilite,satisfaction_rh
0,E001,2019,Industrie,Occitanie,2018,1,529,115830.2,0.1706,7.7
1,E001,2020,Industrie,Occitanie,2018,1,590,125110.5,0.1497,6.7
2,E001,2021,Industrie,Occitanie,2018,1,610,123760.0,0.1594,5.5
3,E001,2022,Industrie,Occitanie,2018,1,652,127441.6,0.1703,8.2
4,E001,2023,Industrie,Occitanie,2018,1,731,140627.0,0.0057,7.0


En pratique, on combine généralement les deux mécaniques en une seule ligne :


In [20]:
entreprises_teletravail = df[df["teletravail"] == 1]
print(f"{len(entreprises_teletravail)} lignes sur {len(df)} concernent des entreprises en télétravail")


232 lignes sur 404 concernent des entreprises en télétravail


### 2.5. Combiner plusieurs conditions

Pour combiner plusieurs conditions dans Pandas, on utilise les symboles :

- `&` pour "et" : les deux conditions doivent être vraies en même temps ;
- `|` pour "ou" : au moins une des conditions doit être vraie.

> [!NOTE]
> On n'utilise pas les opérateurs logiques `and` ("et") et `or` ("ou") de Python natif pour combiner des conditions portant sur des Series Pandas. De plus, chaque condition doit être placée entre parenthèses. Elles permettent à Pandas d'évaluer correctement chaque condition avant de les combiner.


In [21]:
# Entreprises en télétravail ET dans le secteur Technologies

filtre = df[(df["teletravail"] == 1) & (df["secteur"] == "Technologies")]
filtre.head()


,entreprise_id,annee,secteur,region,annee_creation,teletravail,effectifs,chiffre_affaires_keur,rentabilite,satisfaction_rh
5,E002,2019,Technologies,Nouvelle-Aquitaine,2009,1,381,75355.0,0.1907,8.0
6,E002,2020,Technologies,Nouvelle-Aquitaine,2009,1,379,72634.8,0.1020,5.4
7,E002,2021,Technologies,Nouvelle-Aquitaine,2009,1,378,77716.4,0.1693,8.0
8,E002,2022,Technologies,Nouvelle-Aquitaine,2009,1,346,83106.0,0.1043,7.7
9,E002,2023,Technologies,Nouvelle-Aquitaine,2009,1,384,73128.4,0.0645,7.9


> [!IMPORTANT]
> Il faut faire attention à ne pas oublier les parenthèses autour de chaque condition. En effetn, écrire `df[df["teletravail"] == 1 & df["secteur"] == "Technologies"]` (sans les parenthèses) provoque une erreur, à cause de l'ordre de priorité des opérateurs en Python. Prenez le réflexe de **toujours mettre entre paranthèses chaque condition** quand vous en combinez plusieurs.

### 2.6. Trier les données

Il peut être aussi utile d'ordonner les données (après sélection et filtrage) selon une ou plusieurs variables. Avec `pandas`, le tri permet donc de réorganiser les lignes d'un DataFrame selon les valeurs d'une ou plusieurs colonnes. À cet effet, on utilise principalement la méthode `sort_values()`.


In [22]:
# Les 5 entreprises (lignes) avec le chiffre d'affaires le plus élevé

df.sort_values("chiffre_affaires_keur", ascending=False).head()


,entreprise_id,annee,secteur,region,annee_creation,teletravail,effectifs,chiffre_affaires_keur,rentabilite,satisfaction_rh
4,E001,2023,Industrie,Occitanie,2018,1,731,140627.0,0.0057,7.0
3,E001,2022,Industrie,Occitanie,2018,1,652,127441.6,0.1703,8.2
1,E001,2020,Industrie,Occitanie,2018,1,590,125110.5,0.1497,6.7
2,E001,2021,Industrie,Occitanie,2018,1,610,123760.0,0.1594,5.5
190,E038,2023,Santé,Île-de-France,1993,1,499,121937.5,0.0357,7.7


**Décomposition de la commande** 

- `df.sort_values(...)` : Trie les lignes du DataFrame `df` en fonction d'une colonne spécifique.
- `"chiffre_affaires_keur"` : Désigne la colonne utilisée pour le tri (le chiffre d'affaires en milliers d'euros).
- `ascending=False` : Active le tri décroissant (du plus grand chiffre d'affaires au plus petit).
- `.head()` : Conserve uniquement les premières lignes du résultat (5 par défaut).

**À vous de jouer** : construisez un DataFrame `df_grandes_2023`, contenant uniquement les lignes pour lesquelles `annee` vaut 2023 **et** `effectifs` est supérieur à 200. Combien de lignes obtenez-vous ?


In [23]:
# Votre code ici


## 3. Nettoyer les données

Un jeu de données réel n'est presque jamais parfaitement propre. Avant toute analyse, il faut systématiquement vérifier : valeurs manquantes, doublons, types incohérents. On procède donc au nettoyage de données qui consiste à **identifier et corriger les problèmes qui peuvent affecter la qualité ou l'analyse des données**. Il s'agit de préparer les données pour qu'elles puissent être analysées de manière fiable.

Pour cela, on va commencer par vérifier les données avec `df.info()` pour connaître les types de variables et repérer certaines valeurs manquantes. On peut également rechercher les doublons avec `df.duplicated()` et compter les valeurs manquantes avec `df.isna().sum()`.

### 3.1. Repérer les valeurs manquantes

Une valeur manquante est une donnée qui n'a pas été renseignée ou qui est absente. La méthode `.isna()` permet de repérer ces valeurs. Elle retourne `True` lorsqu'une valeur est manquante et `False` lorsqu'elle est présente. Par exemple, si notre DataFrame contient :

```
    age   salaire
0    25    150000
1   NaN     200000
2    32       NaN
```

`df.isna()` retourne :

```
     age  salaire
0  False    False
1   True    False
2  False     True
```

> [!NOTE]
> NaN (Not a Number) est l'une des représentations utilisées par Pandas pour les valeurs manquantes. Selon le contexte, une donnée manquante peut également être représentée par None ou pd.NA.

On peut ensuite utiliser `.sum()` pour compter les valeurs manquantes dans chaque colonne.

In [24]:
# Nombre de valeurs manquantes, par colonne

df.isna().sum()


entreprise_id            0
annee                    0
secteur                  0
region                   0
annee_creation           0
teletravail              0
effectifs                0
chiffre_affaires_keur    0
rentabilite              0
satisfaction_rh          8
dtype: int64

**Comment lire ce résultat ?** `.isna()` donne un tableau de `True`/`False` de la même taille que `df`. En lui appliquant `.sum()`, on additionne ces `True`/`False` colonne par colonne et comme `True` vaut `1` et `False` vaut `0` en Python, cette somme donne directement le **nombre** de valeurs manquantes par colonne. C'est une astuce de Pandas qu'il faut bien comprendre : `.sum()` sur des booléens = un comptage.

On voit ici que `satisfaction_rh` a quelques valeurs manquantes (8). C'est cohérent avec ce que `.info()` nous montrait déjà à la section 1.

### 3.2. Que faire des valeurs manquantes ?

Après avoir repéré les valeurs manquantes, il faut décider comment les traiter. Il n'existe pas de règle universelle ; le choix dépend de **la nature des données, de la quantité de valeurs manquantes et de l'objectif de l'analyse**.

Deux approches sont généralement utilisées.

#### 3.2.1. Option 1 : supprimer les observations concernées

La méthode `.dropna()` permet de supprimer les lignes contenant des valeurs manquantes. Avec `subset`, on peut préciser la colonne à prendre en compte :

In [25]:
# Option 1 : supprimer les lignes qui ont une valeur manquante dans une colonne donnée

df_sans_na = df.dropna(subset=["satisfaction_rh"])
print(f"{len(df)} lignes avant, {len(df_sans_na)} lignes après suppression")


404 lignes avant, 396 lignes après suppression


Ici, seules les lignes pour lesquelles satisfaction_rh est manquante sont supprimées.

> Cependant, supprimer des observations signifie aussi perdre de l'information. Si les valeurs manquantes sont liées aux caractéristiques des répondants ou des entreprises, cette suppression peut introduire un biais dans les résultats.

#### 3.2.2. Option 2 : remplacer les valeurs manquantes

Une autre possibilité consiste à remplacer les valeurs manquantes par une valeur estimée ou calculée. Par exemple, on peut remplacer les valeurs manquantes par la moyenne de la colonne :

In [26]:
# Option 2 : remplacer les valeurs manquantes par une valeur (ex. la moyenne de la colonne)

moyenne_satisfaction = df["satisfaction_rh"].mean()

df_rempli = df.copy()  # on travaille sur une copie pour ne pas modifier df directement

df_rempli["satisfaction_rh"] = (
    df_rempli["satisfaction_rh"].fillna(moyenne_satisfaction)  # fillna() remplace ici chaque valeur manquante par la moyenne calculée.
)

print(f"Valeurs manquantes restantes : {df_rempli['satisfaction_rh'].isna().sum()}")  # Pour vérifier qu'il ne reste plus de valeurs manquantes :


Valeurs manquantes restantes : 0


**Comment choisir entre les deux ?**

- **Supprimer** (`dropna`) est simple et sûr si vous avez très peu de valeurs manquantes et beaucoup de données, mais vous perdez de l'information. Et si les valeurs manquantes ne sont pas "au hasard" (ex. les entreprises les moins satisfaites répondent moins souvent), vous introduisez un biais.
- **Remplacer** (`fillna`) conserve toutes vos lignes, mais vous "inventez" une valeur. À utiliser avec prudence, en étant transparent sur ce choix méthodologique dans votre recherche.

> [!NOTE]
> Traiter les valeurs manquantes est un choix méthodologique, pas seulement une opération technique. Le chercheur doit comprendre pourquoi des données sont manquantes avant de décider de les supprimer ou de les remplacer.

*Pourquoi `.copy()` à la ligne 3 ?* Sans `.copy()`, `df_rempli` ne serait qu'une autre façon de désigner les mêmes données que `df` en mémoire (ce qu'on appelle une "vue") ; et les modifier via `df_rempli` **modifierait aussi `df`** sans que vous vous en rendiez compte. `.copy()` crée une véritable copie indépendante.

### 3.3. Repérer et traiter les doublons

Un doublon est **une observation qui apparaît plusieurs fois** dans un jeu de données. Dans Pandas, la méthode `.duplicated()` permet de repérer les lignes qui sont des copies exactes d'une ligne déjà présente.

#### 3.3.1. Compter les doublons

Pour connaître le nombre de lignes en double :


In [27]:
# Nombre de lignes strictement identiques à une autre ligne

df.duplicated().sum()  # df.duplicated() examine chaque ligne et retourne True lorsqu'elle est identique à une ligne précédente. Et avec .sum(), on compte les True.


np.int64(4)

Cela signifie ici que **4 lignes** sont des doublons d'une ligne précédente.

> [!IMPORTANT]
> `.duplicated()` considère ici qu'une ligne est un doublon lorsque toutes ses colonnes sont identiques à celles d'une ligne précédente.

#### 3.3.2. Voir les doublons

Pour afficher les lignes identifiées comme doublons :

In [28]:
# Voir les doublons

df[df.duplicated()]


,entreprise_id,annee,secteur,region,annee_creation,teletravail,effectifs,chiffre_affaires_keur,rentabilite,satisfaction_rh
34,E007,2022,Industrie,Île-de-France,1999,0,218,34612.8,0.1017,8.3
211,E042,2023,Technologies,Hauts-de-France,2016,1,193,16940.4,0.1284,6.6
213,E043,2019,Technologies,Hauts-de-France,2004,0,85,9943.2,0.0910,7.8
284,E057,2019,Services aux entreprises,Hauts-de-France,1992,1,91,20628.1,0.0837,5.8


Cette instruction utilise le résultat de `df.duplicated()` comme filtre pour ne conserver que les lignes pour lesquelles le résultat est `True`.

#### 3.3.3. Supprimer les doublons

La méthode `.drop_duplicates()` permet de supprimer les lignes en double :

In [29]:
# Supprimer les doublons

df_sans_doublons = df.drop_duplicates()  # Par défaut, Pandas conserve la première occurrence et supprime les suivantes.

print(f"{len(df)} lignes avant, {len(df_sans_doublons)} lignes après suppression des doublons") # Vérifier le nombre de lignes avant et après


404 lignes avant, 400 lignes après suppression des doublons


Cela signifie ici que 4 lignes en double ont été supprimées.

> **Attention aux faux doublons**. Deux observations ne sont pas nécessairement des doublons simplement parce qu'elles ont certaines valeurs identiques. Il est donc important de comprendre ce qui définit une observation unique dans vos données avant de supprimer des doublons.

**À vous de jouer** : Nous avons vu comment supprimer les doublons et comment traiter les valeurs manquantes. Combien de lignes reste-t-il si vous enchaînez `drop_duplicates()` **puis** `dropna(subset=["satisfaction_rh"])` sur `df` ?

> [!TIP]
> Avec Pandas, les méthodes peuvent être enchaînées les unes à la suite des autres en utilisant un point `.`.


In [30]:
# Votre code ici


### 3.4. Vérifier les types

Avant d'analyser une colonne, il est important de vérifier que son type correspond bien à la nature des données qu'elle contient. Par exemple, une variable représentant le nombre d'employés devrait être numérique :

```
effectifs
25
42
18
63
```

Mais il peut arriver qu'une colonne soit enregistrée comme du texte (`str`) à cause du fichier source ou de certaines valeurs inattendues. Par exemple :

```
effectifs
25
42
"18 "
63
```

L'espace après 18 peut causer des problèmes lors de l'importation ou du traitement des données. Dans ces cas, `.describe()` ne fonctionnera pas comme attendu sur cette colonne, et les calculs (additions, moyennes...) échoueront ou donneront des résultats absurdes.

Pour cela, il faut toujours faire une vérification avec la méthode `.info()` (vu en section 1) pour connaître le type de chaque colonne. Ensuite, lorsque nous savons qu'une colonne doit être numérique, nous pouvons utiliser la méthode `pd.to_numeric()`.

Exemple illustratif : Notre colonne est déjà propre. Le code suivant ne fera donc rien de visible ici. Cependant, c'est le réflexe à avoir face à un vrai fichier où ce ne serait pas le cas.


In [31]:
df["effectifs"] = pd.to_numeric(
    df["effectifs"], errors="coerce"
)

Cette instruction demande à Pandas de convertir les valeurs de `effectifs` en nombres.

**Que signifie `errors="coerce"` ?**

Certaines valeurs peuvent être impossibles à convertir. Si on a par exemple :

```
25
42
NC
63
```

Avec l'instruction :

```python
pd.to_numeric(df["effectifs"], errors="coerce")
```

Pandas va convertir les valeurs numériques normalement et transforme les valeurs impossibles à convertir en NaN :

```
25
42
NaN
63
```

Ainsi, le programme ne s'arrête pas à cause d'une valeur problématique.

> [!IMPORTANT]
> Attention, `errors="coerce"` ne corrige pas la donnée. Il transforme simplement les valeurs impossibles à convertir en valeurs manquantes. Il faut donc vérifier ensuite combien de NaN ont été créés, avec : `df["effectifs"].isna().sum()`. Si le résultat est, par exemple, `1`, cela signifie qu'une valeur n'a pas pu être convertie.

***Retenez bien, qu'avant de calculer une moyenne, une somme ou de réaliser une analyse statistique, il faut s'assurer que les variables ont le bon type.***

## 4. Créer et transformer des variables

En analyse de données, il est fréquent de devoir **créer de nouvelles variables ou transformer celles qui existent déjà**. Par exemple, à partir du chiffre d'affaires et du nombre de salariés, nous pouvons calculer un indicateur de productivité. Nous pouvons également transformer une variable numérique en catégorie ou regrouper plusieurs modalités d'une variable. Avec Pandas, ces opérations peuvent généralement être réalisées directement sur les colonnes d'un DataFrame.

### 4.1. Créer une colonne calculée

Pour créer une nouvelle colonne, on l'assigne comme s'il s'agissait d'une variable Python classique ; sauf qu'ici, l'opération s'applique **automatiquement à toutes les lignes en une fois** (on parle d'opération "vectorisée") :


In [32]:

df["ca_par_salarie"] = df["chiffre_affaires_keur"] / df["effectifs"]          # Ici, nous créons la colonne ca_par_salarie en divisant le chiffre d'affaires par le nombre de salariés.

# On peut ensuite vérifier le résultat
df[["entreprise_id", "annee", "chiffre_affaires_keur", "effectifs", "ca_par_salarie"]].head()


,entreprise_id,annee,chiffre_affaires_keur,effectifs,ca_par_salarie
0,E001,2019,115830.2,529,218.960681
1,E001,2020,125110.5,590,212.051695
2,E001,2021,123760.0,610,202.885246
3,E001,2022,127441.6,652,195.462577
4,E001,2023,140627.0,731,192.376197


**Qu'est-ce qu'une opération vectorisée ?**

Avec l'instruction `df["chiffre_affaires_keur"] / df["effectifs"]`, Pandas effectue automatiquement la division pour chaque ligne correspondante des deux colonnes. Si le DataFrame contient 404 lignes, Pandas effectue donc le calcul pour les 404 lignes sans que nous ayons besoin d'écrire une boucle `for`. On parle d'opération vectorisée.

Avec une boucle, il faudrait parcourir les lignes une par une. Avec Pandas, on peut généralement effectuer directement l'opération sur les colonnes :

```python
df["colonne_a"] / df["colonne_b"]
```

C'est l'un des grands avantages de Pandas. Le code est alors plus court, plus lisible et plus efficace que lorsqu'on traite manuellement chaque ligne.

### 4.2. Utiliser  `apply()` pour des transformations personnalisées

Toutes les transformations ne se résument pas à une addition, une soustraction ou une division. Supposons que nous souhaitions classer les entreprises selon leur nombre de salariés :

Pour des transformations plus complexes qu'une simple opération mathématique, `apply()` permet d'exécuter une fonction personnalisée sur chaque ligne (ou chaque valeur) d'une colonne.

- moins de 50 → Petite ;
- de 50 à 249 → Moyenne ;
- 250 ou plus → Grande.

Nous pouvons d'abord créer une fonction :


In [33]:
def categoriser_taille(effectif):
    if effectif < 50:
        return "Petite"
    elif effectif < 250:
        return "Moyenne"
    else:
        return "Grande"

df["taille_entreprise"] = df["effectifs"].apply(categoriser_taille)
df[["entreprise_id", "effectifs", "taille_entreprise"]].head()


,entreprise_id,effectifs,taille_entreprise
0,E001,529,Grande
1,E001,590,Grande
2,E001,610,Grande
3,E001,652,Grande
4,E001,731,Grande


Cette fonction est celle que nous avons apprise au module 01 (sections fonctions). Elle reçoit une valeur puis retourne un résultat. Nous pouvons ensuite l'appliquer à toute la colonne avec `apply()` :

`df["taille_entreprise"] = df["effectifs"].apply(categoriser_taille)`

Pandas prend chaque valeur de `effectifs`, la transmet à `categoriser_taille()` et place le résultat dans la nouvelle colonne. On peut vérifier le résultat avec :

```python
df[
    ["entreprise_id", "effectifs", "taille_entreprise"]
].head()
```

Comme : « Applique la fonction categoriser_taille à chaque valeur de la colonne `effectifs`. »

`apply()` est particulièrement utile lorsque la transformation nécessite une logique personnalisée qui ne peut pas être exprimée simplement avec une opération vectorisée.


### 4.3. Recoder des catégories avec `map()`

En recherche, il est également fréquent de devoir recoder une variable catégorielle. Par exemple, nous pouvons souhaiter regrouper plusieurs régions en deux grandes zones pour simplifier l'analyse. Dans ce cas, on peut écrire le code suivant :

In [34]:

zones = {
    "Île-de-France": "Zone A",
    "Auvergne-Rhône-Alpes": "Zone A",
    "Nouvelle-Aquitaine": "Zone B",
    "Hauts-de-France": "Zone B",
    "Occitanie": "Zone B",
}

df["zone"] = df["region"].map(zones)

df[["region", "zone"]].drop_duplicates() # Vérifier les correspondances


,region,zone
0,Occitanie,Zone B
5,Nouvelle-Aquitaine,Zone B
10,Île-de-France,Zone A
20,Hauts-de-France,Zone B
36,Auvergne-Rhône-Alpes,Zone A


Ici, `zones` est un dictionnaire Python. Chaque région constitue une clé et la zone correspondante constitue sa valeur. Nous pouvons appliquer cette correspondance à la colonne region avec `map()`. Ainsi, `map()` prend un dictionnaire et remplace chaque valeur de la colonne par sa correspondance dans le dictionnaire ; exactement la logique des dictionnaires vue au module 01, appliquée ici à une colonne entière.

Cette opération permet notamment de **recoder des modalités, regrouper des catégories ou simplifier une variable pour une analyse**.

### 4.4. `apply()` ou `map()` ?

Retenez simplement ceci :

- opération mathématique simple → opération vectorisée ;
- transformation nécessitant une logique personnalisée → apply() ;
- correspondance entre des catégories et de nouvelles valeurs → map().

Il existe souvent plusieurs façons d'effectuer une même transformation dans Pandas. L'objectif n'est pas de mémoriser toutes les méthodes, mais de comprendre quelle approche correspond au problème que l'on cherche à résoudre.

**À vous de jouer** : créez une colonne `rentabilite_pct`, qui exprime `rentabilite` en pourcentage plutôt qu'en proportion (c'est-à-dire multipliée par 100). Affichez les 5 premières lignes des colonnes `rentabilite` et `rentabilite_pct` pour vérifier votre résultat.


In [35]:
# Votre code ici


## 5. Grouper et agréger avec `groupby`

En recherche, on cherche généralement à comparer des groupes d'observations. Par exemple :

- les entreprises qui pratiquent le télétravail et celles qui ne le pratiquent pas ;
- les entreprises de différents secteurs ;
- les entreprises selon leur taille ;
- les répondants selon leur catégorie d'âge.

Pandas permet de réaliser ce type de comparaison avec la méthode `groupby()`.

C'est ici que notre question de recherche commence vraiment à prendre forme : **`groupby` est l'outil qui permet de comparer des groupes entre eux** ; exactement ce dont on a besoin pour comparer les entreprises en télétravail et les autres.

> [!NOTE]
> `groupby()` permet de regrouper les observations selon une ou plusieurs variables, puis d'effectuer un calcul séparément pour chaque groupe.

### 5.1. Le principe : "diviser, appliquer, combiner"

`groupby()` fonctionne toujours selon le même principe en trois étapes, même si on l'écrit en une seule ligne de code :

1. **Diviser** les données en groupes (ex. un groupe par valeur de `teletravail`)
2. **Appliquer** un calcul à chaque groupe séparément (ex. la moyenne)
3. **Combiner** les résultats dans un nouveau tableau, un résultat par groupe

Imaginons que notre DataFrame contienne une variable teletravail indiquant si une entreprise pratique ou non le télétravail. Nous pouvons calculer le chiffre d'affaires moyen selon le télétravail :

In [36]:

df.groupby("teletravail")["chiffre_affaires_keur"].mean()


teletravail
0    41522.774419
1    40836.421121
Name: chiffre_affaires_keur, dtype: float64

**Décomposons cette instruction**

Pandas :

- sépare `df` en deux groupes (les lignes où `teletravail` vaut 0, celles où il vaut 1) ;
- calcule la moyenne de `chiffre_affaires_keur` séparément dans chacun des deux groupes ;
- puis affiche les deux résultats côte à côte, avec `.mean()`.

Nous obtenons donc le chiffre d'affaires moyen pour les entreprises sans télétravail (0) et avec télétravail (1).

Cela permet déjà de faire une première comparaison descriptive. Mais attention ! Cette comparaison ne permet pas à elle seule de conclure que le télétravail contribue à une augmentation du chiffre d'affaires. Une différence de moyenne peut avoir de nombreuses autres explications. Les tests statistiques et les modèles étudiés plus loin permettront de vérifier cela.

### 5.2. Grouper selon plusieurs variables

Il est possible de créer des groupes à partir de plusieurs variables. Par exemple, nous pouvons comparer la rentabilité selon le secteur et la pratique du télétravail :


In [37]:
df.groupby(["secteur", "teletravail"])["rentabilite"].mean()


secteur                   teletravail
Commerce                  0              0.070460
                          1              0.097751
Industrie                 0              0.084454
                          1              0.107968
Santé                     0              0.060920
                          1              0.093588
Services aux entreprises  0              0.076086
                          1              0.102531
Technologies              0              0.063431
                          1              0.091686
Name: rentabilite, dtype: float64

Pandas crée alors un groupe pour chaque combinaison de `secteur` et `teletravail` ; chaque ligne correspondant à une combinaison de catégories.

### 5.3. Calculer plusieurs statistiques avec  `.agg() `

Nous ne sommes pas obligés de calculer uniquement une moyenne. La méthode `.agg()` permet d'appliquer plusieurs fonctions d'agrégation à une variable :


In [38]:
df.groupby(["secteur", "teletravail"])["rentabilite"].agg(["mean", "std", "count"])


mean       std  count
secteur                  teletravail                           
Commerce                 0            0.070460  0.047537     20
                         1            0.097751  0.058440     35
Industrie                0            0.084454  0.053099     26
                         1            0.107968  0.075344     50
Santé                    0            0.060920  0.069465     25
                         1            0.093588  0.062218     40
Services aux entreprises 0            0.076086  0.067510     50
                         1            0.102531  0.064048     51
Technologies             0            0.063431  0.056395     51
                         1            0.091686  0.055275     56

Nous obtenons notamment :

- `mean` → la moyenne ;
- `std` → l'écart-type, qui mesure la dispersion des valeurs autour de la moyenne ;
- `count` → le nombre d'observations du groupe.

Le nombre d'observations est particulièrement important. Par exemple, une moyenne calculée sur :

```
Groupe A → 150 observations
Groupe B → 4 observations
```

ne doit pas être interprétée de la même manière.


### 5.4. Présenter les résultats avec `pivot_table()`

La méthode `groupby()` produit parfois un résultat difficile à lire lorsque plusieurs variables sont utilisées. Dans ces cas, Pandas propose la méthode `pivot_table()` pour présenter certaines analyses sous la forme d'un tableau croisé. Par exemple :


In [39]:
pd.pivot_table(
    df,
    values="satisfaction_rh",
    index="secteur",                  # les lignes du tableau croisé
    columns="teletravail",        # les colonnes du tableau croisé
    aggfunc="mean",
)


teletravail,0,1
secteur,,
Commerce,6.870000,6.768571
Industrie,6.772000,7.091837
Santé,6.784000,7.162500
Services aux entreprises,6.427083,6.976000
Technologies,6.312500,6.989286


Dans le code :

- `values="satisfaction_rh"` → la variable que nous voulons analyser ;
- `index="secteur"` → les secteurs apparaissent en lignes ;
- `columns="teletravail"` → les catégories de télétravail apparaissent en colonnes ;
- `aggfunc="mean"` → nous calculons la moyenne.

On obtient alors un vrai tableau croisé secteur × télétravail, avec la satisfaction RH moyenne dans chaque case. On peut ensuite comparer rapidement la satisfaction moyenne selon le secteur et la pratique du télétravail.

**À vous de jouer** : calculez l'effectif moyen (`effectifs`) par `secteur`, trié du secteur avec le plus grand effectif moyen au plus petit. (Astuce : vous pouvez enchaîner `.groupby(...)`, `[...]`, `.mean()` puis `.sort_values(...)`.)


In [40]:
# Votre code ici


## 6. Croiser plusieurs jeux de données

En recherche, les données dont nous avons besoin sont rarement réunies dans un seul fichier. Par exemple, nous pouvons disposer de :

- données financières dans un fichier ;
- données RH dans un autre ;
- informations géographiques dans un troisième ;
- données collectées à plusieurs années dans plusieurs fichiers.

Il faut alors pouvoir réunir ces données. Pandas propose principalement deux outils pour cela :

- `merge()` qui permet de croiser des données à partir d'une variable commune ;
- `concat()` qui permet d'assembler plusieurs DataFrames en les empilant.

### 6.1. Croiser des données avec `merge()`

Pour illustrer le principe, imaginons que nous disposions de deux DataFrames. Le premier contient nos données principales ; par exemple :

```
entreprise_id    chiffre_affaires
E001             1500
E002             2300
E003             1800
E004             2450
```

Le second contient des informations sur le siège social ; par exemple :

```
entreprise_id    ville_siege
E001             Toulouse
E002             Lyon
E003             Bordeaux
E004             Lille
```

Les deux tableaux ont une colonne commune : `entreprise_id`. Cette colonne permet d'identifier les entreprises présentes dans les deux jeux de données. Nous pouvons créer le second DataFrame ainsi :

In [41]:
sieges_sociaux = pd.DataFrame({
    "entreprise_id": ["E001", "E002", "E003", "E004"],
    "ville_siege": ["Toulouse", "Lyon", "Bordeaux", "Lille"],
})

sieges_sociaux


,entreprise_id,ville_siege
0,E001,Toulouse
1,E002,Lyon
2,E003,Bordeaux
3,E004,Lille


Pour croiser les deux DataFrames, nous allons utiliser `merge()` :

`merge()` croise deux DataFrames sur une colonne commune (ici, `entreprise_id`), un peu comme une jointure SQL si vous connaissez déjà ce vocabulaire ; sinon, ne vous en inquiétez pas, l'important est de comprendre le principe ci-dessous.


In [42]:
df_avec_siege = df.merge(
    sieges_sociaux,
    on="entreprise_id",                                    # Le paramètre on indique la colonne qui permet de faire correspondre les deux DataFrames.
    how="left"
)

df_avec_siege[["entreprise_id", "annee", "secteur", "ville_siege"]].head(8)  # Pour vérifier le résultat


,entreprise_id,annee,secteur,ville_siege
0,E001,2019,Industrie,Toulouse
1,E001,2020,Industrie,Toulouse
2,E001,2021,Industrie,Toulouse
3,E001,2022,Industrie,Toulouse
4,E001,2023,Industrie,Toulouse
5,E002,2019,Technologies,Lyon
6,E002,2020,Technologies,Lyon
7,E002,2021,Technologies,Lyon


**Le paramètre `how`**

Le paramètre how indique quelles lignes doivent être conservées lorsqu'une correspondance existe ou non entre les deux DataFrames. Les quatre principales options sont :

| `how=` | Comportement |
|---|---|
| `"left"` | Garde **toutes** les lignes du DataFrame de gauche (`df`), même celles sans correspondance dans `sieges_sociaux` (les colonnes ajoutées seront alors `NaN`) |
| `"right"` | Symétrique : garde toutes les lignes du DataFrame de droite |
| `"inner"` | Garde **seulement** les lignes présentes des deux côtés |
| `"outer"` | Garde **toutes** les lignes des deux côtés, `NaN` là où il n'y a pas de correspondance |

Dans notre exemple, `sieges_sociaux` ne contient que 4 entreprises sur les 80 du jeu de données complet. Avec `how="left"`, on garde donc bien toutes les lignes de `df`. Si une entreprise n'existe pas dans `sieges_sociaux`, Pandas ne peut pas trouver de ville et indique une valeur manquante : `NaN`.

On peut ensuite vérifier combien de correspondances n'ont pas été trouvées :


In [43]:

df_avec_siege["ville_siege"].isna().sum()


np.int64(384)

### 6.2. Pourquoi vérifier le résultat d'un `merge()` ?

Un `merge()` peut modifier le nombre de lignes du DataFrame. Il faudra donc toujours vérifier le résultat d'une jointure.


In [44]:

print("Avant :", len(df))
print("Après :", len(df_avec_siege))


Avant : 404
Après : 404


Si le nombre de lignes a changé alors que vous ne vous y attendiez pas, il faut chercher pourquoi. Cela peut être lié à des identifiants absents ou à des valeurs répétées dans la colonne utilisée pour la jointure.


### 6.3. Empiler des données avec `concat()`

`merge()` permet de croiser des colonnes provenant de différents jeux de données. Mais que faire lorsque nous avons plusieurs DataFrames qui ont les mêmes colonnes et que nous souhaitons simplement les réunir ? C'est le rôle de `concat()`.

Imaginons que nous ayons un DataFrame pour 2019 et un autre pour 2020. Nous pouvons les empiler avec ce code :


In [45]:
df_2019 = df[df["annee"] == 2019]
df_2020 = df[df["annee"] == 2020]

df_2019_2020 = pd.concat([df_2019, df_2020])    # Les lignes du deuxième DataFrame sont ajoutées à la suite de celles du premier.

print(f"{len(df_2019)} + {len(df_2020)} = {len(df_2019_2020)} lignes après concat")   # Vérifier le résultat


82 + 80 = 162 lignes après concat


### 6.4. `merge()` ou `concat()` ?

Une distinction simple permet de retenir l'essentiel :

`merge()` → **je veux croiser des informations**

```
Base financière        Base RH
      ↓                    ↓
      └──── entreprise_id ─┘
               ↓
        Base enrichie
```

`concat()` → **je veux empiler des données**

```
Données 2019
     ↓
Données 2020
     ↓
Données 2019 + 2020
```

> [!NOTE]
> Retene donc qu'il faut utiliser `merge()` lorsque plusieurs jeux de données doivent être reliés grâce à une clé commune. Utilisez `concat()` lorsque vous souhaitez principalement réunir des DataFrames ayant une structure similaire. En recherche, cette distinction est importante, car la manière dont les jeux de données sont assemblés peut avoir un impact sur l'échantillon analysé.


## 7. Gérer les dates

Les dates sont très fréquentes dans les données de recherche : date d'une enquête, date de création d'une entreprise, date d'une transaction, année d'observation, date de publication d'un document, etc. Pour pouvoir manipuler correctement des dates dans Pandas, il faut que la colonne soit reconnue comme une date et non comme du simple texte.

Notre jeu de données contient une variable `annee`, qui indique l'année d'observation :


In [46]:
df["annee"].head()

0    2019
1    2020
2    2021
3    2022
4    2023
Name: annee, dtype: int64

Cette variable est actuellement un entier (`int64`), et non une date. Cela convient pour certaines analyses, mais certaines opérations nécessitent de travailler avec de véritables dates :

- extraire un mois ou un jour ;
- calculer une durée entre deux dates ;
- comparer des dates, etc.

Pour illustrer ces opérations, nous allons créer une date à partir de `annee`.

### 7.1. Créer une date à partir d'une année

Imaginons que nous souhaitions représenter le début de chaque année d'observation. Nous pouvons créer une chaîne de caractères contenant le 1er janvier de chaque année :

In [47]:
dates_texte = df["annee"].astype(str) + "-01-01"

dates_texte.head()

0    2019-01-01
1    2020-01-01
2    2021-01-01
3    2022-01-01
4    2023-01-01
Name: annee, dtype: str

Pour l'instant, ces valeurs sont encore du texte. Nous allons donc les convertir en véritables dates avec `pd.to_datetime()`.

### 7.2. Convertir en `datetime`

`pd.to_datetime()` convertit une colonne de texte en un vrai type "date", ce qui débloque tout un ensemble d'opérations :


In [48]:

df["date_observation"] = pd.to_datetime(dates_texte)

df[["annee", "date_observation"]].head()    # Vérifier le résultat


,annee,date_observation
0,2019,2019-01-01
1,2020,2020-01-01
2,2021,2021-01-01
3,2022,2022-01-01
4,2023,2023-01-01


In [49]:
df[["date_observation"]].info()  # Vérifier le type de colonne

<class 'pandas.DataFrame'>
RangeIndex: 404 entries, 0 to 403
Data columns (total 1 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   date_observation  404 non-null    datetime64[us]
dtypes: datetime64[us](1)
memory usage: 3.3 KB


On obtient `datetime64[us]`. Cela signifie que Pandas reconnaît désormais `date_observation` comme une date, et non comme du texte.

### 7.3. Extraire des composantes d'une date

Une fois la colonne au format `datetime`, nous pouvons utiliser `.dt` pour accéder à différentes composantes de la date. Par exemple : 

In [50]:
df["annee_observation"] = df["date_observation"].dt.year     # Pour récupérer l'année

df["mois_observation"] = df["date_observation"].dt.month   # Pour récupérer le mois

df["jour_observation"] = df["date_observation"].dt.day          # Pour récupérer le jour

# Afficher les résultats

df[["date_observation", "annee_observation", "mois_observation", "jour_observation"]].head()

,date_observation,annee_observation,mois_observation,jour_observation
0,2019-01-01,2019,1,1
1,2020-01-01,2020,1,1
2,2021-01-01,2021,1,1
3,2022-01-01,2022,1,1
4,2023-01-01,2023,1,1


### 7.4. Calculer une durée

L'un des intérêts importants des dates est de pouvoir calculer des écarts de temps. Par exemple, nous pouvons calculer combien d'années séparent l'année de création de l'entreprise de l'année d'observation :

In [51]:

df["age_entreprise"] = (
    df["annee"] - df["annee_creation"]
)


Ici, nous n'avons même pas besoin de `datetime`, car nous travaillons simplement avec deux années. Par exemple :

```
annee_creation    annee    age_entreprise
2018              2019     1
2018              2020     2
2018              2021     3
```

Cette nouvelle variable peut être particulièrement intéressante pour notre analyse : l'âge de l'entreprise pourrait constituer un facteur à prendre en compte lorsque nous comparons les entreprises.

### 7.5. Trouver la période couverte par les données

Nous pouvons également utiliser `.min()` et `.max()` sur `annee` pour connaître la période couverte par notre jeu de données :

In [52]:
df["annee"].min()

np.int64(2019)

In [53]:
df["annee"].max()

np.int64(2023)

Cela permet de déterminer rapidement la première et la dernière année disponibles. Nous pouvons faire la même chose avec notre colonne `date_observation` :


In [54]:
df["date_observation"].min()

Timestamp('2019-01-01 00:00:00')

In [55]:
df["date_observation"].max()

Timestamp('2023-01-01 00:00:00')

**À vous de jouer**

- Quelle est la première année présente dans le jeu de données ?
- Quelle est la dernière année ?
- Calculez l'âge moyen des entreprises à partir de la variable `age_entreprise`.

> [!TIP]
> Indice : utilisez .min(), .max() et .mean().


In [56]:
# Votre code ici


## 8. Retour sur notre question de recherche

Nous pouvons maintenant revenir à notre question de recherche de départ :

> **Est-ce que le télétravail contribue à une meilleure performance des entreprises ?**

Notre jeu de données contient plusieurs indicateurs qui peuvent nous permettre d'examiner cette question :

- teletravail → pratique ou non du télétravail ;
- chiffre_affaires_keur → chiffre d'affaires ;
- effectifs → nombre de salariés ;
- rentabilite → rentabilité ;
- satisfaction_rh → satisfaction RH.

Nous pouvons construire une première synthèse descriptive :

In [57]:
df["ca_par_salarie"] = (
    df["chiffre_affaires_keur"] / df["effectifs"]
)

synthese = df.groupby("teletravail").agg(
    ca_par_salarie_moyen=("ca_par_salarie", "mean"),
    rentabilite_moyenne=("rentabilite", "mean"),
    satisfaction_rh_moyenne=("satisfaction_rh", "mean")
)

synthese


,ca_par_salarie_moyen,rentabilite_moyenne,satisfaction_rh_moyenne
teletravail,,,
0,149.834806,0.070740,6.553012
1,151.887980,0.098822,7.004783


Ce tableau nous permet de décrire les différences entre les entreprises qui pratiquent le télétravail et celles qui ne le pratiquent pas.

**Mais attention !** Nous ne pouvons pas encore conclure que le télétravail cause une meilleure performance. Une différence de rentabilité moyenne peut, par exemple, s'expliquer par le secteur d'activité, la taille de l'entreprise, son âge ou d'autres caractéristiques. C'est justement ce qui fera la transition vers les statistiques. Pandas nous permet de **préparer et d'explorer les données**. Les statistiques nous permettront ensuite **de déterminer si les différences observées sont suffisamment importantes pour être considérées comme statistiquement significatives**.

**Ce que nous savons maintenant faire :**

- charger des données ;
- explorer un DataFrame ;
- sélectionner et filtrer des observations ;
- nettoyer les données ;
- créer et transformer des variables ;
- regrouper et agréger les observations ;
- croiser plusieurs jeux de données ;
- manipuler des dates ;
- construire de premières statistiques descriptives.

C'est **la base du travail empirique en sciences de gestion**. Avant de tester une hypothèse, il faut être capable de préparer, comprendre et transformer correctement ses données.

## À vous de jouer

4 exercices :

- **Exercice 1** : chargement et exploration d'un jeu de données réel
- **Exercice 2** : nettoyage (valeurs manquantes, doublons, types)
- **Exercice 3** : `groupby` et statistiques par groupe
- **Exercice 4** : croiser deux jeux de données avec `merge`

Corrigés disponibles dans [`solutions/02-pandas-donnees/`](../../solutions/02-pandas-donnees/). Essayez de résoudre les exercices d'abord vous-même, avant de consulter les solutions.


## Suite

Une fois à l'aise avec `pandas`, passez au module [03-visualisation](../03-visualisation/) pour apprendre à représenter graphiquement vos données.
